chalk-engine/src/logic.rs

```diff
@@ -946,6 +954,14 @@ impl<'forest, C: Context + 'forest, CO: ContextOps<C> + 'forest> SolveState<'for
                 }
             }
             None => {
+                if floundered {
+                    // The strand floundered when trying to select a subgoal.
+                    // This will always return a `RootSearchFail`, either because the
+                    // root table floundered or we yield with `QuantumExceeded`.
+                    return NoRemainingSubgoalsResult::RootSearchFail(
```

Comment:
I wonder if we actually need this. I think if we mark the table as floundered in `pursue_answer` and just return `QuantumExceeded`, the correct logic will be followed the next loop around to actually handle that.

At least, that would be cleaner.

Reply:
I think the logic that would handle this is here: https://github.com/rust-lang/chalk/blob/master/chalk-engine/src/logic.rs#L1274

Reply:
I think I handled this (moved the call to `on_subgoal_selection_flounder` into `pursue_answer_from_floundered`), but I didn't totally follow your second comment - the link is just to the doc before `pursue_answer`. Could you explain it a bit more?

Reply:
Oops, master has changed since I made that comment, so line number is different. Here is a permalink version: https://github.com/rust-lang/chalk/blob/1f9c529f771083794a536f6fb2b1325119f8177a/chalk-engine/src/logic.rs#L1250

Reply:
As far as your change, I think all you really need to add is:
```
// This table is marked as floundered
let table = self.stack.top().table;
debug!("Marking table {:?} as floundered!", table);
self.forest.tables[table].mark_floundered();
```
instead of calling `on_subgoal_selection_flounder`

However, it's interesting that you didn't add it to the "non trivial but new answer" case. And that still works. Makes me wonder if there's a hole in our testing.

Reply:
Hmm... I tried this out and it seems to cause a bunch of other test failures. It seems like a bunch are because `select_subgoal` may now return `SubgoalSelection::NotSelected` (previously `Floundered`) even when there are remaining subgoals.